source/common/stats/lazy_init.h

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include "source/common/common/thread.h"
+
+namespace Envoy {
+namespace Stats {
+
+template <typename StatsStructType> class DirectStats;
+template <typename StatsStructType> class LazyInit;
+
+/**
+ * Interface for stats lazy initialization.
+ * To reduce memory and CPU consumption, Envoy can enable the bootstrap config
+ * :ref:`enable_lazyinit_stats
+ * <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.enable_lazyinit_stats>`.
+ * A 'StatsStructType' is only created when any of its field is referenced.
+ * See more context: https://github.com/envoyproxy/envoy/issues/23575
+ */
+template <typename StatsStructType> class LazyCompatibleInterface {
+public:
+  // Helper function to get-or-create and return the StatsStructType object.
+  virtual StatsStructType* instantiate() = 0;
+
+  virtual ~LazyCompatibleInterface() = default;
+};
+
+/**
+ * Lazy-initialization wrapper for StatsStructType, intended for deferred instantiation of a block
+ * of stats that might not be needed in a given Envoy process.
+ *
+ * This class is thread-safe -- instantiations can occur on multiple concurrent threads.
+ * This is used when
+ * :ref:`enable_lazyinit_stats
+ * <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.enable_lazyinit_stats>` is enabled.
+ */
+template <typename StatsStructType>
+class LazyInit : public LazyCompatibleInterface<StatsStructType> {
+public:
+  // Capture the stat names object and the scope with a ctor, that can be used to instantiate a
+  // StatsStructType object later.
+  // Caller should make sure scope and stat_names outlive this object.
+  LazyInit(const typename StatsStructType::StatNameType& stat_names, Stats::ScopeSharedPtr scope)
+      : initialized_([&scope]() -> Gauge& {
+          Stats::StatNamePool pool(scope->symbolTable());
+          return Stats::Utility::gaugeFromElements(
+              *scope, {pool.add(StatsStructType::typeName()), pool.add("initialized")},
+              Stats::Gauge::ImportMode::Accumulate);
+        }()),
+        ctor_([scope = std::move(scope), &stat_names, this]() -> StatsStructType* {
+          initialized_.inc();
+          return new StatsStructType(stat_names, *scope);
+        }) {
+    if (initialized_.value() > 0) {
```

Comment:
I should've thought have this before. Reading a stat from production code is an anti-pattern that makes it difficult to disable stats via config without breaking production.

See https://github.com/envoyproxy/envoy/issues/14610

This is not the only case, so I don't know if this should block this PR, but it should definitely get a TODO and xref that issue number.

This may be kind of a corner case, but we should think about what happens in the code if we disable stats by prefix -- think that will work out OK but we should discuss. the NullGauge implementation will always return 0 for 'initialized', which I think will give us acceptable behavior because none of the stats will be retained on xDS update, but they will be disabled anyway. but we should probably test for that.

If we do something like disabling `\.initialized` then the stats will work, but they will be dropped on xDS update. This would be pretty subtle issue, but may be unlikely to occur, I'm not sure. We should test for that case too.

Reply:
I knew that bug. I am not sure we should worry that here. 
we can possibly make the suffix really "unreadable" if that helps to stop accidental suffix match? something like "initialized!$"?

Reply:
let's not do that. That'll just stress the ability of stats sinks to be able to parse, or force Envoy to escape them.

This is the place to worry about it. I don't view it as blocking is this is off by default but I think this needs to be considered, and made invisible to users. It shouldn't be difficult; just mark with a bit (there are spare bits in the stats flags) to make admin and sinks skip it.